File: internal/search/query.go
// Search operators: the small grammar the search box accepts. One parser
// lives here and is the only authority on what a term means; the server
// maps the parsed terms onto the index, and the web client parses the
// same grammar (web/src/opsearch.ts) to highlight operators as chips,
// offer completions, and filter the switcher. Anything the grammar does
// not recognise is plain full text — an unknown operator is searched,
// never rejected.
package search

import (
	"strings"
	"time"
)

// Operator names.
const (
	OpTag      = "tag"
	OpPath     = "path"
	OpSpace    = "space"
	OpIs       = "is"
	OpHas      = "has"
	OpAuthor   = "author"
	OpBefore   = "before"
	OpAfter    = "after"
	dateFormat = "2006-01-02"
)

// isValues are the is: selectors.
var isValues = map[string]bool{
	"untagged": true,
	"task":     true,
	"html":     true,
}

// hasValues are the has: selectors.
var hasValues = map[string]bool{
	"image":      true,
	"attachment": true,
}

// Term is one token of a query. Op is empty for plain text; Value holds
// the operator's argument, Text the plain text of a word or phrase.
type Term struct {
	// Raw is the term exactly as typed, negation mark included.
	Raw string
	// Negated says the term excludes what it matches (-tag:x).
	Negated bool
	// Op is the operator name (tag, path, …) or "" for plain text.
	Op string
	// Value is the operator's argument.
	Value string
	// Date is Value parsed, for before:/after: only.
	Date time.Time
	// Quoted says a text term was written in double quotes (an exact
	// phrase) — or that an operator's value was quoted.
	Quoted bool
	// Text is the searchable text of a plain term ("" for operators).
	Text string
}

// Query is a parsed search query.
type Query struct {
	Terms []Term
}

// TextTerms returns the plain text terms, positive and negated apart.
// Words too short for the trigram index ride along; the caller decides
// how to run them.
func (q Query) TextTerms() (positive, negative []Term) {
	for _, t := range q.Terms {
		if t.Op == "" && t.Text != "" {
			if t.Negated {
				negative = append(negative, t)
			} else {
				positive = append(positive, t)
			}
		}
	}
	return positive, negative
}

// HasFilters reports whether any operator term is present.
func (q Query) HasFilters() bool {
	for _, t := range q.Terms {
		if t.Op != "" {
			return true
		}
	}
	return false
}

// Parse splits a query into terms. Whitespace separates terms; a term is
// an operator (name:value, the value optionally double-quoted), a bare
// #tag, a double-quoted exact phrase, or a plain word. A leading -
// negates the term that follows it. A name that is not a known operator,
// an is:/has: selector that does not exist, or a date that is not
// YYYY-MM-DD demotes the whole term to plain text.
func Parse(q string) Query {
	var terms []Term
	rs := []rune(strings.TrimSpace(q))
	i := 0
	for i < len(rs) {
		for i < len(rs) && isSpace(rs[i]) {
			i++
		}
		if i >= len(rs) {
			break
		}
		start := i
		negated := false
		if rs[i] == '-' && i+1 < len(rs) && !isSpace(rs[i+1]) {
			negated = true
			i++
		}
		if i >= len(rs) {
			break
		}
		switch {
		case rs[i] == '"':
			content, end := quotedRun(rs, i)
			terms = append(terms, Term{
				Raw: string(rs[start:end]), Negated: negated,
				Quoted: true, Text: content,
			})
			i = end
		case rs[i] == '#' && i+1 < len(rs) && !isSpace(rs[i+1]):
			j := wordEnd(rs, i+1)
			terms = append(terms, Term{
				Raw: string(rs[start:j]), Negated: negated,
				Op: OpTag, Value: strings.ToLower(string(rs[i+1 : j])),
			})
			i = j
		default:
			name, opLen := opName(rs, i)
			if opLen > 0 {
				valStart := i + opLen
				if valStart < len(rs) && rs[valStart] == '"' {
					v, end := quotedRun(rs, valStart)
					terms = append(terms, operatorTerm(rs, start, valStart, end, negated, name, v, true))
					i = end
				} else {
					j := wordEnd(rs, valStart)
					v := string(rs[valStart:j])
					terms = append(terms, operatorTerm(rs, start, valStart, j, negated, name, v, false))
					i = j
				}
			} else {
				j := wordEnd(rs, i)
				terms = append(terms, Term{
					Raw: string(rs[start:j]), Negated: negated,
					Text: string(rs[i:j]),
				})
				i = j
			}
		}
	}
	return Query{Terms: terms}
}

// operatorTerm builds an operator term, or a plain-text term when the
// operator or its value does not hold up. valStart is the index of the
// value's first rune (or its opening quote); rawEnd is the term's last
// index, both computed by Parse.
func operatorTerm(rs []rune, start, valStart, rawEnd int, negated bool, name, value string, quoted bool) Term {
	demote := func() Term {
		return Term{Raw: string(rs[start:rawEnd]), Negated: negated, Text: string(rs[start+negLen(negated) : rawEnd])}
	}
	switch name {
	case OpIs:
		if !isValues[strings.ToLower(value)] {
			return demote()
		}
		value = strings.ToLower(value)
	case OpHas:
		if !hasValues[strings.ToLower(value)] {
			return demote()
		}
		value = strings.ToLower(value)
	case OpBefore, OpAfter:
		d, err := time.Parse(dateFormat, value)
		if err != nil {
			return demote()
		}
		return Term{Raw: string(rs[start:rawEnd]), Negated: negated, Op: name, Value: value, Date: d.UTC(), Quoted: quoted}
	case OpTag:
		if value == "" {
			return demote()
		}
		value = strings.ToLower(value)
	case OpPath, OpSpace, OpAuthor:
		if value == "" {
			return demote()
		}
	default:
		return demote()
	}
	return Term{Raw: string(rs[start:rawEnd]), Negated: negated, Op: name, Value: value, Quoted: quoted}
}

func negLen(negated bool) int {
	if negated {
		return 1
	}
	return 0
}

// quotedRun reads the quoted run starting at the quote rs[open]: it
// returns the content between the quotes and the index just past the
// closing quote, or the end of the string when it never closes.
func quotedRun(rs []rune, open int) (string, int) {
	for i := open + 1; i < len(rs); i++ {
		if rs[i] == '"' {
			return string(rs[open+1 : i]), i + 1
		}
	}
	return string(rs[open+1:]), len(rs)
}

// wordEnd returns the index of the first space at or after i.
func wordEnd(rs []rune, i int) int {
	for i < len(rs) && !isSpace(rs[i]) {
		i++
	}
	return i
}

// opName reads a known lowercase operator name followed by a colon at
// i, returning the name and its length including the colon; 0 when the
// run of letters is not a known operator.
func opName(rs []rune, i int) (string, int) {
	j := i
	for j < len(rs) && rs[j] >= 'a' && rs[j] <= 'z' {
		j++
	}
	if j == i || j >= len(rs) || rs[j] != ':' {
		return "", 0
	}
	name := string(rs[i:j])
	switch name {
	case OpTag, OpPath, OpSpace, OpIs, OpHas, OpAuthor, OpBefore, OpAfter:
		return name, j - i + 1
	}
	return "", 0
}

func isSpace(r rune) bool {
	return r == ' ' || r == '\t' || r == '\n' || r == '\r'
}

File: android/app/src/main/java/com/collinpendleton/yana/data/search/Query.kt
package com.collinpendleton.yana.data.search

import java.time.LocalDate
import java.time.ZoneOffset

/**
 * The search box grammar, a line-for-line port of the server's parser
 * (internal/search/query.go) so the same box contents mean the same
 * thing online and offline: whitespace-separated terms; a term is an
 * operator (name:value, the value optionally quoted), a bare #tag, a
 * quoted phrase, or plain text; a leading - negates. A name that is not
 * a known operator, an is:/has: selector that does not exist, or a date
 * that is not YYYY-MM-DD demotes the term to plain text — anything the
 * grammar does not recognise is searched, never rejected.
 */

/** Operator names. */
object Ops {
    const val TAG = "tag"
    const val PATH = "path"
    const val SPACE = "space"
    const val IS = "is"
    const val HAS = "has"
    const val AUTHOR = "author"
    const val BEFORE = "before"
    const val AFTER = "after"
}

private val isValues = setOf("untagged", "task", "html")
private val hasValues = setOf("image", "attachment")

/** One token of a query: plain text, or an operator and its argument. */
data class Term(
    /** The term exactly as typed, negation mark included. */
    val raw: String,
    /** The term excludes what it matches (-tag:x). */
    val negated: Boolean,
    /** The operator name, or "" for plain text. */
    val op: String = "",
    /** The operator's argument. */
    val value: String = "",
    /** [value] parsed, for before:/after: only: epoch nanoseconds. */
    val date: Long = 0,
    /** The text term was written in double quotes (an exact phrase). */
    val quoted: Boolean = false,
    /** The searchable text of a plain term ("" for operators). */
    val text: String = "",
)

/** A parsed search query. */
data class SearchQuery(val terms: List<Term>) {
    /**
     * The plain text terms, positive and negated apart. Words too short
     * for the trigram index ride along; the caller decides how to run
     * them.
     */
    fun textTerms(): Pair<List<Term>, List<Term>> {
        val positive = terms.filter { it.op == "" && it.text.isNotEmpty() && !it.negated }
        val negative = terms.filter { it.op == "" && it.text.isNotEmpty() && it.negated }
        return positive to negative
    }
}

/** Splits a query into terms, exactly as the server does. */
fun parseQuery(q: String): SearchQuery {
    val terms = ArrayList<Term>()
    val rs = q.trim().toCodePoints()
    var i = 0
    while (i < rs.size) {
        while (i < rs.size && isSpace(rs[i])) i++
        if (i >= rs.size) break
        val start = i
        var negated = false
        if (rs[i] == '-'.code && i + 1 < rs.size && !isSpace(rs[i + 1])) {
            negated = true
            i++
        }
        if (i >= rs.size) break
        when {
            rs[i] == '"'.code -> {
                val (content, end) = quotedRun(rs, i)
                terms += Term(raw = substring(rs, start, end), negated = negated, quoted = true, text = content)
                i = end
            }
            rs[i] == '#'.code && i + 1 < rs.size && !isSpace(rs[i + 1]) -> {
                val j = wordEnd(rs, i + 1)
                terms += Term(
                    raw = substring(rs, start, j),
                    negated = negated,
                    op = Ops.TAG,
                    value = substring(rs, i + 1, j).lowercase(),
                )
                i = j
            }
            else -> {
                val (name, opLen) = opName(rs, i)
                if (opLen > 0) {
                    val valStart = i + opLen
                    if (valStart < rs.size && rs[valStart] == '"'.code) {
                        val (v, end) = quotedRun(rs, valStart)
                        terms += operatorTerm(rs, start, valStart, end, negated, name, v, quoted = true)
                        i = end
                    } else {
                        val j = wordEnd(rs, valStart)
                        terms += operatorTerm(rs, start, valStart, j, negated, name, substring(rs, valStart, j), quoted = false)
                        i = j
                    }
                } else {
                    val j = wordEnd(rs, i)
                    terms += Term(raw = substring(rs, start, j), negated = negated, text = substring(rs, i, j))
                    i = j
                }
            }
        }
    }
    return SearchQuery(terms)
}

/**
 * Builds an operator term, or a plain-text term when the operator or its
 * value does not hold up — the server's demotion rules.
 */
private fun operatorTerm(
    rs: IntArray,
    start: Int,
    valStart: Int,
    rawEnd: Int,
    negated: Boolean,
    name: String,
    value: String,
    quoted: Boolean,
): Term {
    fun demote() = Term(
        raw = substring(rs, start, rawEnd),
        negated = negated,
        text = substring(rs, start + if (negated) 1 else 0, rawEnd),
    )
    return when (name) {
        Ops.IS -> if (!isValues.contains(value.lowercase())) {
            demote()
        } else {
            Term(raw = substring(rs, start, rawEnd), negated = negated, op = name, value = value.lowercase(), quoted = quoted)
        }
        Ops.HAS -> if (!hasValues.contains(value.lowercase())) {
            demote()
        } else {
            Term(raw = substring(rs, start, rawEnd), negated = negated, op = name, value = value.lowercase(), quoted = quoted)
        }
        Ops.BEFORE, Ops.AFTER -> {
            val day = runCatching { LocalDate.parse(value) }.getOrNull()
            if (day == null) demote()
            else Term(raw = substring(rs, start, rawEnd), negated = negated, op = name, value = value, date = day.atStartOfDay(ZoneOffset.UTC).toEpochSecond() * 1_000_000_000L, quoted = quoted)
        }
        Ops.TAG -> if (value.isEmpty()) {
            demote()
        } else {
            Term(raw = substring(rs, start, rawEnd), negated = negated, op = name, value = value.lowercase(), quoted = quoted)
        }
        Ops.PATH, Ops.SPACE, Ops.AUTHOR -> if (value.isEmpty()) {
            demote()
        } else {
            Term(raw = substring(rs, start, rawEnd), negated = negated, op = name, value = value, quoted = quoted)
        }
        else -> demote()
    }
}

/** Reads the quoted run starting at the quote: content, and index past the close. */
private fun quotedRun(rs: IntArray, open: Int): Pair<String, Int> {
    var i = open + 1
    while (i < rs.size) {
        if (rs[i] == '"'.code) return substring(rs, open + 1, i) to i + 1
        i++
    }
    return substring(rs, open + 1, rs.size) to rs.size
}

/** The index of the first space at or after i. */
private fun wordEnd(rs: IntArray, i: Int): Int {
    var j = i
    while (j < rs.size && !isSpace(rs[j])) j++
    return j
}

/** A known lowercase operator name followed by a colon, or "" with 0. */
private fun opName(rs: IntArray, i: Int): Pair<String, Int> {
    var j = i
    while (j < rs.size && rs[j] in 'a'.code..'z'.code) j++
    if (j == i || j >= rs.size || rs[j] != ':'.code) return "" to 0
    return when (val name = substring(rs, i, j)) {
        Ops.TAG, Ops.PATH, Ops.SPACE, Ops.IS, Ops.HAS, Ops.AUTHOR, Ops.BEFORE, Ops.AFTER -> name to j - i + 1
        else -> "" to 0
    }
}

private fun isSpace(r: Int): Boolean =
    r == ' '.code || r == '\t'.code || r == '\n'.code || r == '\r'.code

/** The string's code points, the unit the server's rune-based parser works in. */
private fun String.toCodePoints(): IntArray {
    val out = IntArray(length)
    var n = 0
    var i = 0
    while (i < length) {
        val cp = codePointAt(i)
        out[n++] = cp
        i += Character.charCount(cp)
    }
    return out.copyOf(n)
}

/** The substring between two code point indices. */
private fun substring(rs: IntArray, from: Int, to: Int): String {
    val sb = StringBuilder()
    for (k in from until to) sb.appendCodePoint(rs[k])
    return sb.toString()
}
